### PR TITLE
(HC-11) Port includer functionality

### PR DESCRIPTION
**Pull request by whopper**
These commits add all of the necessary bits for the includer.

### CODE REVIEW

File: lib/src/simple_includer.cc [hunk 1]
```diff
@@ -42,4 +164,32 @@ namespace hocon {
...
+        return _delegate->include(move(context), move(what));
+    }
+
+    shared_object proxy::include_file(shared_include_context context, std::string what) const {
+        if (auto del = dynamic_pointer_cast<const config_includer_file>(_delegate)) {
+            return del->include_file(move(context), move(what));
+        } else {
+            // TODO: where does this come from?
+            // return include_file_without_fallback(move(context), move(what));
+            throw config_exception("Not sure where include_file_without_fallback is supposed to come from");
```
Comment (whopper) [inline lib/src/simple_includer.cc]: TODO: figure this out

File: lib/src/values/simple_config_object.cc [hunk 2]
```diff
@@ -264,6 +264,24 @@ namespace hocon {
...
 
+    shared_ptr<simple_config_object> simple_config_object::empty() {
+        return empty_instance();
+    }
+
+    shared_ptr<simple_config_object> simple_config_object::empty(shared_origin origin) {
+        if (origin == nullptr) {
+            return empty();
+        } else {
+            // TODO: do this correctly
```
Comment (whopper) [inline lib/src/values/simple_config_object.cc]: TODO: fix this as well

File: lib/src/config_parser.cc [hunk 3]
```diff
@@ -70,6 +72,53 @@ namespace hocon { namespace config_parser {
...
+        // See https://github.com/typesafehub/config/issues/160
+        if (array_count > 0 && obj->get_resolve_status() != resolve_status::RESOLVED) {
+            // TODO: use the right exception - parseError in Java
+            throw config_exception("Due to current limitations of the config parser, when an include statement is nested inside a list value,\n"
+                                           "${} substitutions inside the included file cannot be resolved correctly. Either move the include outside of the list value or\n"
+                                           "remove the ${} statements from the included file.");
+        }
+
+        /*
+        if (!_path_stack.empty()) {
```
Comment (whopper) [inline lib/src/config_parser.cc]: TODO: fix this

**Comment by MikaelSmith** [comment]
Failed cpplint.

File: lib/src/values/simple_config_object.cc (hunk 2 above)
Comment (MikaelSmith) [inline lib/src/values/simple_config_object.cc]: If it's wrong, we should probably throw an exception. Or is it currently used in tests?

Comment (whopper) [inline lib/src/values/simple_config_object.cc]: @MikaelSmith I forgot to put WIP on this but I'm gonna be going through some of these today, will decided what to do with this.

**Comment by whopper** [comment]
Still todo: remove references to PROPERTIES, finish includer tests.

**Pull request merged**